WMFData/Sources/WMFData/Data Controllers/Shared/WMFPageViewsDataController.swift

```diff
@@ -178,4 +202,45 @@ public final class WMFPageViewsDataController {
         
         return results
     }
+    
+    public func fetchPageViewDates(startDate: Date, endDate: Date, moc: NSManagedObjectContext? = nil) throws -> [WMFPageViewDay] {
+        let context: NSManagedObjectContext
+        if let moc {
+            context = moc
+        } else {
+            context = try coreDataStore.viewContext
+        }
+        
+        let results: [WMFPageViewDay] = try context.performAndWait {
+            let predicate = NSPredicate(format: "timestamp >= %@ && timestamp <= %@", startDate as CVarArg, endDate as CVarArg)
+            
+            let pageViewsDict = try self.coreDataStore.fetchGrouped(
+                entityType: CDPageView.self,
+                predicate: predicate,
+                propertyToCount: "timestamp",
+                propertiesToGroupBy: ["timestamp"],
+                propertiesToFetch: ["timestamp"],
+                in: context
+            )
+            
+            var pageViewDays: [WMFPageViewDay] = []
+            
+            for dict in pageViewsDict {
+                guard let timestamp = dict["timestamp"] as? Date,
+                      let count = dict["count"] as? Int else {
+                    continue
+                }
+                
+                let dayFormatter = Date.FormatStyle()
+                    .weekday(.wide)
+                let dayOfWeek = timestamp.formatted(dayFormatter)
+                
+                pageViewDays.append(WMFPageViewDay(day: dayOfWeek, viewCount: count))
+            }
+            
+            return pageViewDays
```

Comment:
I don't think this will work the way we want - what the existing `fetchPageViewCounts` method does above is take all CDPageViews, group them by _page_, and count up how many are in each group. So you end up with a [page: count] dictionary like:

[{"page": "Cat", "count": 1},
{"page": "Dog", "count": 3},
{"page": "Bird", "count": 2}]*

_*The "page" value type here is actually an `NSManagedObjectID` type, but I'm using the article title here for clarity._

Meaning Cat was viewed once, Dog was viewed 3 times, etc.

But what you're doing here is similar but you're grouping by timestamp. Since timestamp is hyper-specific, you're likely not going to get more than one count per grouping, since I don't expect it's possible to view multiple articles at the exact same time:

[{"timestamp": "2024-10-09T14:20:58Z", "count": 1},
{"timestamp": "2024-10-09T14:21:32Z", "count": 1},
{"timestamp": "2024-10-09T14:22:46Z", "count": 1}]

Meaning the number of articles viewed at exactly `2024-10-09T14:20:58Z` was one, etc.

So I think going the `fetchedGrouped` route doesn't make sense here. What I suggest doing is pulling an array of CDPageViews from Core Data, then looping through and manually adding them to your own counts dictionary, like this:

```
            let predicate = NSPredicate(format: "timestamp >= %@ && timestamp <= %@", startDate as CVarArg, endDate as CVarArg)
            
            guard let cdPageViews = try self.coreDataStore.fetch(entityType: CDPageView.self, predicate: predicate, fetchLimit: nil, in: context) else {
                return []
            }
            
            // [dayofWeek: count]
            var countsDictionary: [String: Int] = [:]
            
            for cdPageView in cdPageViews {
                
                // Figure out cdPageView's dayOfWeek here from cdPageView.timestamp
                
                // Add or increment counts dictionary here
            }
            
            // turn counts dictionary into array of WMFPageViewDay, return that
```

Reply:
Done!